provider/core/src/error.rs

```diff
@@ -263,3 +265,28 @@ impl From<std::io::Error> for DataError {
         DataErrorKind::Io(e.kind()).into_error()
     }
 }
+
+/// A [`DataError`] that also returns a [`DataPayload`] to the caller.
+#[allow(clippy::exhaustive_structs)] // this type is stable
+pub struct DataErrorWithPayload<M: DataMarker>(pub DataPayload<M>, pub DataError);
```

Comment:
Why move this here? This is only used by the converter thing, which itself is only used for the zero-copy test. I don't think we should make this available outside `datagen`.

Reply:
It seems like a generally useful type; if we added some function that needed to use this type outside of `icu_provider::datagen` (which could be either inside or outside of `icu_provider`), we would need to either add a similar duplicated type, point into `icu_provider::datagen` to get it at that point, or move the type and re-export it from `icu_provider::datagen`.

Reply:
> ... or move the type and re-export it from icu_provider::datagen.

That's a non-breaking change which we can do if we ever need that. I'm not convinced that it's a generally useful type, I find it quite confusing actually.

Reply:
Okay, it seems you feel more strongly on this than I. I'll move it back to datagen; do you have an opinion on the name `DataErrorWithPayload` versus the original name `ReturnedPayloadError`? (I prefer `DataErrorWithPayload`)

Reply:
No strong opinion, I'd also be happy to not name this and just use a pair.

Reply:
I think it has a name so we can implement the `?` operator for it

But I'd be okay changing the small number of call sites to use `map_err`. Seems like that may be the easiest solution.

Reply:
You should be able to define `From<(DataError, DataPayload<M>)> for DataError`. But I'd prefer using `map_err` over this.

Reply:
Removed. Turns out the `?` appears to never be used anywhere on this error?